lib/common/xml.c

```diff
@@ -602,19 +602,37 @@ pcmk__xe_remove_matching_attrs(xmlNode *element,
     }
 }
 
+/*!
+ * \internal
+ * \brief Create a new XML element under a given parent
+ *
+ * \param[in,out] parent  XML element that will be the new element's parent
+ *                        (\c NULL to create a new XML document with the new
+ *                        node as root)
+ * \param[in]     name    Name of new element
+ *
+ * \return Newly created XML element, or \c NULL on memory allocation failure
+ */
 xmlNode *
-create_xml_node(xmlNode * parent, const char *name)
-{
-    xmlDoc *doc = NULL;
+pcmk__xe_create(xmlNode *parent, const char *name)
+{
+    /* @TODO Either assert on memory allocation failure here, or DON'T assert
+     * on it in pcmk__xml_copy().
```

Comment:
I agree. To your usual point, it's not really appropriate to dump core in these cases. However, I definitely agree that it's a ton of clutter to error-check this stuff. And we already have the precedent of GLib functions and `pcmk__str_update()` that assert on malloc failure.

Is there a reason why we don't fast-exit (without dumping core) in cases like this and `pcmk__str_update()`?

Reply:
That would be better for memory failures. Maybe:

    #define PCMK__MEM_ASSERT(ptr) do { if (ptr == NULL) { \
        crm_abort(__FILE__, __func__, __LINE__, "Out of memory", FALSE, FALSE) } while (0)

Reply:
Yeah I like that. Taking a boolean and expecting `ptr != NULL` to be the condition is *arguably* more readable, since asserts traditionally take a boolean, but this is more concise so I don't care either way.

logging_internal.h?

Reply:
Do you prefer all caps? The other internal logging macros are lowercase

Reply:
Could optionally have a `pcmk__calloc()` that calls `calloc()` followed by `pcmk__mem_assert()`. Bare calls to `pcmk__mem_assert()` could come after library allocators, specifically thinking of libxml2.

Reply:
`logging_internal.h` (emphasizing the message) or `results_internal.h` (emphasizing the abort) would do

Lowercase is fine. Uppercase is especially helpful for macros that use arguments more than once, as a warning about side effects, but doesn't really add anything otherwise.

Reply:
I'm thinking `pcmk__assert_alloc()` for the wrapper